### PR TITLE
gh-113280: Always close socket if SSLSocket creation failed

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
* Issue: gh-113280

### CODE REVIEW

File: Lib/ssl.py [hunk 1]
```diff
@@ -1000,65 +1000,61 @@ def _create(cls, sock, server_side=False, do_handshake_on_connect=True,
         )
         self = cls.__new__(cls, **kwargs)
         super(SSLSocket, self).__init__(**kwargs)
-        sock_timeout = sock.gettimeout()
-        sock.detach()
+        try:
+            sock_timeout = sock.gettimeout()
+            sock.detach()
```
Comment (graingert) [inline Lib/ssl.py]: I think it would be better to call `sock.gettimeout()` and `sock.detach()` outside the try/catch - as if either of those raise you could get a double close

```python
        sock_timeout = sock.gettimeout()
        kwargs = dict(
            family=sock.family, type=sock.type, proto=sock.proto,
            fileno=sock.detatch()
        )
        self = cls.__new__(cls, **kwargs)
        super(SSLSocket, self).__init__(**kwargs)
        try:
            self._context = context
            ...
```

Comment (serhiy-storchaka) [reply Lib/ssl.py]: But what if `__new__` or `__init__` raise?

File: Lib/ssl.py [hunk 2]
```diff
@@ -994,71 +994,67 @@ def _create(cls, sock, server_side=False, do_handshake_on_connect=True,
...
-        self._session = session
-        self._closed = False
-        self._sslobj = None
-        self.server_side = server_side
-        self.server_hostname = context._encode_hostname(server_hostname)
-        self.do_handshake_on_connect = do_handshake_on_connect
-        self.suppress_ragged_eofs = suppress_ragged_eofs
-
-        # See if we are connected
+        socket.detach()
```
Comment (graingert) [inline Lib/ssl.py]:
```suggestion
        sock.detach()
```

File: Lib/ssl.py (hunk 1 above)
Comment (graingert) [reply Lib/ssl.py]: Then they should be responsible for closing the passed in fileno

Comment (serhiy-storchaka) [reply Lib/ssl.py]: They are not. If you pass a fileno, you are responsible (in this case it is the responsibility of `sock` until `detach()` is called).

**Comment by serhiy-storchaka** [comment]
I added tests, although the original case reported in the issue is not covered by tests. But I added tests for two cases in which an explicit `close()` was already called and which are now merged with a more general solution. And a new test for incorrect hostname, it emits a resource warning without this fix.
